Python/ceval.c

```diff
@@ -881,6 +881,262 @@ _Py_CheckRecursiveCall(PyThreadState *tstate, const char *where)
     return 0;
 }
 
+
+// PEP 634: Structural Pattern Matching
+
+
+static PyObject*
+match_keys(PyThreadState *tstate, PyObject *map, PyObject *keys)
+{
+    // Return a tuple of values corresponding to keys, with error checks for
+    // duplicate/missing keys.
+    assert(PyTuple_CheckExact(keys));
+    Py_ssize_t nkeys = PyTuple_GET_SIZE(keys);
+    if (!nkeys) {
+        // No keys means no items.
+        return PyTuple_New(0);
+    }
+    // We use the two argument form of map.get(key, default) for two reasons:
+    // - Atomically check for a key and get its value without error handling.
+    // - Don't cause key creation or resizing in dict subclasses like
+    //   collections.defaultdict that define __missing__ (or similar).
+    _Py_IDENTIFIER(get);
+    PyObject *get = _PyObject_GetAttrId(map, &PyId_get);
+    if (!get) {
+        return NULL;
+    }
+    PyObject *seen = NULL;
+    PyObject *values = NULL;
+    // dummy = object()
+    PyObject *dummy = _PyObject_CallNoArg((PyObject *)&PyBaseObject_Type);
+    if (!dummy) {
+        goto fail;
+    }
+    seen = PySet_New(NULL);
+    if (!seen) {
+        goto fail;
+    }
+    values = PyTuple_New(nkeys);
+    if (!values) {
+        goto fail;
+    }
+    for (Py_ssize_t i = 0; i < nkeys; i++) {
+        PyObject *key = PyTuple_GET_ITEM(keys, i);
+        if (PySet_Contains(seen, key) || PySet_Add(seen, key)) {
+            if (!_PyErr_Occurred(tstate)) {
+                // Seen it before!
+                _PyErr_Format(tstate, PyExc_ValueError,
+                              "mapping pattern checks duplicate key (%R)", key);
+            }
+            goto fail;
+        }
+        PyObject *value = PyObject_CallFunctionObjArgs(get, key, dummy, NULL);
+        if (!value || value == dummy) {
+            // Key not in map.
+            Py_XDECREF(value);
+            goto fail;
+        }
+        PyTuple_SET_ITEM(values, i, value);
+    }
+    // Success.
+    Py_DECREF(get);
+    Py_DECREF(dummy);
+    Py_DECREF(seen);
+    return values;
+fail:
+    Py_DECREF(get);
+    Py_XDECREF(dummy);
+    Py_XDECREF(seen);
+    Py_XDECREF(values);
+    return NULL;
+}
+
+static PyObject*
+dict_without_keys(PyThreadState *tstate, PyObject *map, PyObject *keys)
+{
+    // copy = dict(map)
+    // for key in keys:
+    //     del copy[key]
+    // return copy
+    PyObject *copy = PyDict_New();
+    if (!copy || PyDict_Update(copy, map)) {
+        Py_XDECREF(copy);
+        return NULL;
+    }
+    // This may seem a bit inefficient, but keys is rarely big enough to
+    // actually impact runtime.
+    for (Py_ssize_t i = 0; i < PyTuple_GET_SIZE(keys); i++) {
+        if (PyDict_DelItem(copy, PyTuple_GET_ITEM(keys, i))) {
+            Py_DECREF(copy);
+            return NULL;
+        }
+    }
+    return copy;
+}
+
+static int
+get_match_args(PyThreadState *tstate, PyObject *type, PyObject **match_args,
+               int *match_self)
+{
+    // try:
```

Comment:
This comment basically describes the bytecode sequence that should be generated. Why not do this in the compiler?

Reply:
Well, because this isn't *actually* what we're doing here... it's just approximately equivalent Python code to help explain the behavior, which can be a bit subtle. Call it pseudocode, if you want.

Setting up an exception handler and actually performing all of this logic in Pythonland is needlessly complicated for what we're trying to do.

Reply:
It sounds like you are saying that this comment is misleading. Please remove it, or fix it.